src/libsyntax/parse/parser.rs

```diff
@@ -732,6 +732,12 @@ impl<'a> Parser<'a> {
                   format!("expected {} here", expect)))
             };
             let mut err = self.fatal(&msg_exp);
+            if self.token.is_ident_named("and") {
+                err.help("Use `&&` instead of `and` for the boolean operator");
```

Comment:
Could you use the following?

```
err.span_suggestion_with_applicability(
    self.span,
    "use `&&` instead of `and` for the boolean operator",
    "&&".to_string(),
    Applicability::MaybeIncorrect,
);
```

This way the output will be presented inline and third party tools will be able to apply the suggestion in their UIs.

Same change in all four cases.